[MSE] Ensure we are not removing the current time during automatic eviction

calvaris opened the pull request:
#### 6d7afcf0850ffbb4b6f06f3063ca131fbbf281cd
<pre>
[MSE] Ensure we are not removing the current time during automatic eviction
<a href="https://bugs.webkit.org/show_bug.cgi?id=247052">https://bugs.webkit.org/show_bug.cgi?id=247052</a>

Reviewed by Alicia Boya Garcia.

In certain cases, depending on the sync frame positions, we could end up removing the current time, which would lead to
playback stopping. Now we ensure that the previous sync frame to current time is taking into account when calculating
the maximumRangeEnd.

* LayoutTests/media/media-source/media-source-append-buffer-full-evict-prior-to-end-expected.txt: Added.
* LayoutTests/media/media-source/media-source-append-buffer-full-evict-prior-to-end.html: Added.
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::findPreviousSyncSamplePresentationTime):
(WebCore::SourceBufferPrivate::evictCodedFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:

Canonical link: <a href="https://commits.webkit.org/256939@main">https://commits.webkit.org/256939@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14e4a38a91dd5a2f19f288e9fd0943b0eaaf1324

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30440 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106809 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167075 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101259 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6850 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35289 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89684 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103490 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102953 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83919 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32130 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86997 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75056 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/570 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20292 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/554 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21751 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4782 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5355 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1798 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41081 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -357,6 +357,18 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
     return trackBuffer.removeSamples(samples, logPrefix);
 }
 
+MediaTime SourceBufferPrivate::findSyncFrameMediaTimePreviousTo(const MediaTime& time)
```
**Rename frame to sample** in the function name. Personally, I prefer to refer to them as frames, as I often do so in writing, so as to not cause confusion with PCM audio samples. But given that the class is called `MediaSample` and all the existing functions call them samples, we should be consistent -- we don't want anyone to wonder the difference between a sample and a frame in WebKit MSE when they're synonyms.

**Replace MediaTime with PresentationTime** in the function name. The user needs to know which one of the two is returned.

Similarly, the argument should be called `presentationTime`, `time` is ambiguous.

e.g. `findPreviousSyncSamplePresentationTime(const MediaTime& presentationTime)` -- it's a very verbose and very specific name for a very specific function. Similarly. It would benefit from a comment explaining its purpose in prose:

> For this SourceBuffer, given a presentation time anywhere in the presentation timeline, find the presentation time for the sync sample it depends on.
> 
> If this SourceBuffer contains more than one TrackBuffer, the sample with earliest (i.e. furthest apart) presentation time is chosen.
> 
> Erasures on presentation time ranges that are strictly less than contain the returned value are guaranteed to never trigger the deletion of the frame with the provided presentationTime or any of their dependencies. Note any frame dependent on a sync sample must have a presentation time greater or equal to that of the sync sample.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -357,6 +357,18 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
 
+MediaTime SourceBufferPrivate::findSyncFrameMediaTimePreviousTo(const MediaTime& time)
+{
+    MediaTime previousSyncFrameMediaTime = time;
```
`previousSyncSamplePresentationTime`

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -357,6 +357,18 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
+    for (auto& trackBufferKeyValue : m_trackBufferMap) {
+        TrackBuffer& trackBuffer = trackBufferKeyValue.value;
+        auto sampleIterator = trackBuffer.samples().decodeOrder().findSyncSamplePriorToPresentationTime(time);
+        if (sampleIterator != trackBuffer.samples().decodeOrder().rend() && time > sampleIterator->first.second)
```
Save `sampleIterator->first.second` in a variable so that it's more readable. I don't expect most readers to remember that `sampleIterator->first.first` is decodeTime and `sampleIterator->first.second` is presentationTime. I had to look for it in the code because I never remember which one is first and which one is second.

ntrrgc commented on LayoutTests/media/media-source/media-source-append-buffer-full-evict-prior-to-end.html:
```diff
@@ -0,0 +1,87 @@
+        for (var pts = firstPts; pts <= lastPts; pts++) {
+            try {
+                consoleWrite('Appending PTS='+pts);
+                sourceBuffer.appendBuffer(makeASample(pts, pts, 1, 1, 1, pts % 5 == 0 ? SAMPLE_FLAG.SYNC : SAMPLE_FLAG.NONE, 1));
```
This is a nice simple test, but it would also be nice to test our code doesn't break in presence of B-frames. Here you have some example frames with B-frames:
![b-frames](https://user-images.githubusercontent.com/1002436/201664392-bf525a0d-c74f-4304-8324-50cc82e25a41.jpg)

ntrrgc replied:
Addedum: in this example I've given the second I-frame is not an IDR-frame, and therefore should not be considered a sync sample as far as WebKit is concerned. [Quoting the MSE spec](https://www.w3.org/TR/media-source/#random-access-point):

> Random Access Point: A position in a [media segment](https://www.w3.org/TR/media-source/#media-segment) where decoding ***and continuous playback*** can begin without relying on any previous data in the segment. For video this ***tends*** to be the location of I-frames.

The second I-frame is not a random access point/sync sample/IDR frame because, although it is enough to decode anything coming after it in decode order, it's not enough to play everything after it in presentation order. Notably, the B-frame that is immediately after it in presentation order cannot be played without the previous GOP.

ntrrgc commented on Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
```diff
@@ -357,6 +357,21 @@ static PlatformTimeRanges removeSamplesFromTrackBuffer(const DecodeOrderSampleMa
+        if (sampleIterator == trackBuffer.samples().decodeOrder().rend())
+            continue;
+        const MediaTime& samplePresentationTime = sampleIterator->first.second;
+        if (time > samplePresentationTime)
```
Nit (feel free to ignore): This is a classic "find minimum" algorithm. Flipping the condition to `samplePresentationTime < time` feels more natural to me, whereas the other way in first read makes me think it's an unrelated condition. Maybe it's just me though.

ntrrgc reviewed (approved):
LGTM now.

calvaris committed:
[MSE] Ensure we are not removing the current time during automatic eviction
https://bugs.webkit.org/show_bug.cgi?id=247052

Reviewed by Alicia Boya Garcia.

In certain cases, depending on the sync frame positions, we could end up removing the current time, which would lead to
playback stopping. Now we ensure that the previous sync frame to current time is taking into account when calculating
the maximumRangeEnd.

* LayoutTests/media/media-source/media-source-append-buffer-full-evict-prior-to-end-expected.txt: Added.
* LayoutTests/media/media-source/media-source-append-buffer-full-evict-prior-to-end.html: Added.
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::findPreviousSyncSamplePresentationTime):
(WebCore::SourceBufferPrivate::evictCodedFrames):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:

Canonical link: https://commits.webkit.org/256939@main
